src/basic/xattr-util.c

```diff
@@ -20,110 +21,164 @@
 #include "time-util.h"
 #include "xattr-util.h"
 
+/* Use a single cache for all of *xattrat syscalls (added in kernel 6.13) */
+static thread_local bool have_xattrat = true;
+
+static int normalize_and_maybe_pin_inode(
+                int *fd,
+                const char **path,
+                int *at_flags,
+                int *ret_tfd,
+                bool *ret_opath) {
+
+        int r;
+
+        assert(fd);
+        assert(*fd >= 0 || *fd == AT_FDCWD);
+        assert(path);
+        assert(at_flags);
+        assert(ret_tfd);
+        assert(ret_opath);
+
+        if (isempty(*path))
+                *path = NULL; /* Normalize "" to NULL */
+
+        if (*fd == AT_FDCWD) {
+                if (!*path) /* Both unspecified? Then operate on current working directory */
+                        *path = ".";
+
+                *ret_tfd = -EBADF;
+                *ret_opath = false;
+                return 0;
+        }
+
+        *at_flags |= AT_EMPTY_PATH;
+
+        if (!*path) {
+                r = fd_is_opath(*fd);
+                if (r < 0)
+                        return r;
+                *ret_opath = r;
+
+                *ret_tfd = -EBADF;
+                return 0;
+        }
+
+        /* If both have been specified, then we go via O_PATH */
+
+        int tfd = openat(*fd, *path, O_PATH|O_CLOEXEC|(FLAGS_SET(*at_flags, AT_SYMLINK_FOLLOW) ? 0 : O_NOFOLLOW));
```

Comment:
Not your fault, but maybe this does not work when the path points to a symlink without `AT_SYMLINK_FOLLOW`. In that case, an `O_PATH` fd for the symlink is opened, and `getxattr()` with `FORMAT_PROC_FD_PATH()` will be called, and that follows symlink... no? Or do I misunderstand something?

Reply:
`/proc/self/fd/*` are magic links. They always refer to the inode initially pinned via `open()`. No further symlink resolving is done when operating on them.

Reply:
Interesting. I did not know that.